File: Behavioral/Interpreter/CSharp/InterpreterClient.cs
// InterpreterClient.cs

using System;
using System.Collections.Generic;

public class InterpreterClient {
  public static IExpression ParseExpression(string expression) {
    Stack<IExpression> stack = new Stack<IExpression>();

    string[] tokens = expression.Split(' ');
    foreach (string token in tokens) {
      if (token == "+") {
        IExpression right = stack.Pop();
        IExpression left = stack.Pop();
        stack.Push(new AddExpression(left, right));
      } else if (token == "-") {
        IExpression right = stack.Pop();
        IExpression left = stack.Pop();
        stack.Push(new SubtractExpression(left, right));
      } else {
        stack.Push(new NumberExpression(int.Parse(token)));
      }
    }
    return stack.Pop();
  }
}

File: Behavioral/Interpreter/JavaScript/interpreter_client.js
// interpreter_client.js

const NumberExpression = require("./number_expression");
const AddExpression = require("./add_expression");
const SubtractExpression = require("./subtract_expression");

function parseExpression(expression) {
  const stack = [];
  const tokens = expression.split(" ");

  for (const token of tokens) {
    if (token === "+") {
      const right = stack.pop();
      const left = stack.pop();
      stack.push(new AddExpression(left, right));
    } else if (token === "-") {
      const right = stack.pop();
      const left = stack.pop();
      stack.push(new SubtractExpression(left, right));
    } else {
      stack.push(new NumberExpression(parseInt(token, 10)));
    }
  }

  return stack.pop();
}

module.exports = parseExpression;
